include/nlohmann/json.hpp

```diff
@@ -3912,13 +3828,13 @@ class basic_json // NOLINT(cppcoreguidelines-special-member-functions,hicpp-spec
     @sa see @ref operator[](const typename object_t::key_type&) for unchecked
     access by reference
 
-    @since version 1.0.0
+    @since version 1.0.0, KeyType template added in 3.10.0
     */
     // using std::is_convertible in a std::enable_if will fail when using explicit conversions
-    template < class ValueType, typename std::enable_if <
+    template < class KeyType, class ValueType, typename detail::enable_if_t <
                    detail::is_getable<basic_json_t, ValueType>::value
-                   && !std::is_same<value_t, ValueType>::value, int >::type = 0 >
-    ValueType value(const typename object_t::key_type& key, const ValueType& default_value) const
+                   && !std::is_same<value_t, ValueType>::value&& detail::is_usable_as_key_type<basic_json_t, KeyType>::value > ... >
+    ValueType value(const KeyType& key, ValueType&& default_value) const
```

Comment:
When I check out this branch, this is the only function causing compilation problems. The issue here is that when you use `FunctionTemplateParameter &&`, the type deduced is a reference type if you pass an lvalue (a non-temporary). You can resolve this by changing the return type to `std::decay_t<ValueType>` and changing line 3846 to `return it->template get<std::decay_t<ValueType>>();`.

Reply:
Changing line 3846 to

```cpp
return it->template get<typename std::decay<ValueType>::type>();
```

(this should be the C++11 equivalent, right?) yields


```
In file included from ../test/src/unit-regression2.cpp:37:
../include/nlohmann/json.hpp:3846:24: error: non-const lvalue reference to type 'basic_string<...>' cannot bind to a temporary of type 'basic_string<...>'
                return it->template get<typename std::decay<ValueType>::type>();
                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../test/src/unit-regression2.cpp:475:22: note: in instantiation of function template specialization 'nlohmann::basic_json<std::map, std::vector, std::__1::basic_string<char>, bool, long long, unsigned long long, double, std::allocator, adl_serializer, std::__1::vector<unsigned char>>::value<char [2], std::__1::basic_string<char> &>' requested here
        auto val = j.value("x", defval);
                     ^
1 error generated.
```

Reply:
Yes, that's the C++11 equivalent. Did you also apply the transformation to the return type?

Reply:
No, I forgot about that. Now it compiles locally. Let's see...

Reply:
There are still compilation errors in the CI - I will update to the latest `develop` branch to have all changes of the last months in.

Reply:
Ok, now there are several build errors.

```
In file included from ../test/src/unit-element_access2.cpp:32:
../include/nlohmann/json.hpp:3881:16: error: no viable conversion from returned value of type 'typename std::decay<basic_string_view<char>>::type' (aka 'std::__1::basic_string_view<char>') to function return type 'nlohmann::basic_json<>::string_t' (aka 'std::__1::basic_string<char>')
        return value(key, std::string_view(default_value));
               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../test/src/unit-element_access2.cpp:250:29: note: in instantiation of function template specialization 'nlohmann::basic_json<std::map, std::vector, std::__1::basic_string<char>, bool, long long, unsigned long long, double, std::allocator, adl_serializer, std::__1::vector<unsigned char>>::value<char [7]>' requested here
                    CHECK(j.value("string", "bar") == "hello world");
```

```
In file included from ../test/src/unit-json_patch.cpp:32:
../include/nlohmann/json.hpp:3881:16: error: no viable conversion from returned value of type 'typename std::decay<basic_string_view<char>>::type' (aka 'std::__1::basic_string_view<char>') to function return type 'nlohmann::basic_json<>::string_t' (aka 'std::__1::basic_string<char>')
        return value(key, std::string_view(default_value));
               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../test/src/unit-json_patch.cpp:1354:37: note: in instantiation of function template specialization 'nlohmann::basic_json<std::map, std::vector, std::__1::basic_string<char>, bool, long long, unsigned long long, double, std::allocator, adl_serializer, std::__1::vector<unsigned char>>::value<char [8]>' requested here
                INFO_WITH_TEMP(test.value("comment", ""));
```

Any ideas?

Reply:
Well, from a quick look at the code, what is the purpose of this line:

https://github.com/nlohmann/json/blob/575c28ca4c8ffb6571c2378ff32dd025b6b34c79/single_include/nlohmann/json.hpp#L20970-L20974

Do you actually want to return a `string_view`, when the default value is specified as a string litteral in c++17 mode, but return a `string_t` in pre-17 modes? That seems (dangerously) inconsistent to me.

Reply:
That really makes no sense at all. Most likely, I was too aggressively pattern-matching "useless" string constructions.

Reply:
I think the idea came from here: https://github.com/nlohmann/json/pull/2685#discussion_r624746212 (sorry overlooked it when I posted this) and I think  @davidstone has a point in that this should ideally not unconditionally create a temporary `string_t` object if the default isn't used.

However, best I can tell, thats an orthogonal concern that AFAIK already exists in the develop branch (so it isn't a regression in this PR) and can be fixed separately.